src/System.Net.Primitives/src/System/Net/NetworkCredential.cs

```diff
@@ -0,0 +1,220 @@
+            Domain = domain;
+        }
+
+        /// <devdoc>
+        ///    <para>
+        ///       Initializes a new instance of the <see cref='System.Net.NetworkCredential'/>
+        ///       class with name and password set as specified.
+        ///    </para>
+        /// </devdoc>
+        public NetworkCredential(string userName, SecureString password, string domain)
+        {
+            UserName = userName;
+            SecurePassword = password;
+            Domain = domain;
+        }
+
+        /// <devdoc>
+        ///    <para>
+        ///       The user name associated with this credential.
+        ///    </para>
+        /// </devdoc>
+        public string UserName
+        {
+            get
+            {
+                return InternalGetUserName();
+            }
+            set
+            {
+                _userName = value ?? string.Empty;
+            }
+        }
+
+        /// <devdoc>
+        ///    <para>
+        ///       The password for the user name.
+        ///    </para>
+        /// </devdoc>
+        public string Password
+        {
+            get
+            {
+                return InternalGetPassword();
+            }
+            set
+            {
+                _password = SecureStringHelpers.CreateSecureString(value);
+            }
+        }
+
+        /// <devdoc>
+        ///    <para>
+        ///       The password for the user name.
+        ///    </para>
+        /// </devdoc>
+        public SecureString SecurePassword
+        {
+            get
+            {
+                return InternalGetSecurePassword().Copy();
+            }
+            set
+            {
+                _password = value == null ? new SecureString() : value.Copy();
+            }
+        }
+
+        /// <devdoc>
+        ///    <para>
+        ///       The machine name that verifies
+        ///       the credentials. Usually this is the host machine.
+        ///    </para>
+        /// </devdoc>
+        public string Domain
+        {
+            get
+            {
+                return InternalGetDomain();
+            }
+            set
+            {
+                _domain = value ?? string.Empty;
+            }
+        }
+
+        internal string InternalGetUserName()
+        {
+            return _userName;
+        }
+
+        internal string InternalGetPassword()
+        {
+            return SecureStringHelpers.GetPlaintext(_password);
+        }
+
+        internal SecureString InternalGetSecurePassword()
+        {
+            return _password;
+        }
+
+        internal string InternalGetDomain()
+        {
+            return _domain;
+        }
+
+        internal string InternalGetDomainUserName()
+        {
+            string domainUserName = InternalGetDomain();
+            if (domainUserName.Length != 0)
+                domainUserName += "\\";
+            domainUserName += InternalGetUserName();
+            return domainUserName;
```

Comment:
If this were instead written like:
```C#
string domain = InternalGetDomain();
string userName = InternalGetUserName();
return domain.Length != 0 ?
    domain + "\\" + userName :
    userName;
```
I believe that'd save at least one string allocation.